libraries/ms-common/src/can_datagram.c

```diff
@@ -0,0 +1,616 @@
+      return;
+    }
+  }
+  event_raise_no_data(store->repeat_event);
+}
+
+static void prv_process_protocol_version_rx(Fsm *fsm, const Event *e, void *context) {
+  CanDatagramStorage *store = context;
+  CanDatagram *dgram = &store->dgram;
+  // Verify start message received/sent TODO: REMOVE!
+  if (!s_store.start) {
+    LOG_WARN("START MESSAGE NOT RECEIVED. EXITING WITH ERROR...\n");
+    event_raise_no_data(store->error_event);
+    return;
+  }
+  // Check if data has been written yet, if not, re-raise
+  if (!prv_read_rx_buffer(&dgram->protocol_version, PROTOCOL_VERSION_SIZE_BYTES)) {
+    event_raise_no_data(store->repeat_event);
+  } else {
+    if (dgram->protocol_version != CAN_DATAGRAM_VERSION) {
+      event_raise_no_data(store->error_event);
+      return;
+    }
+    event_raise_no_data(store->rx_event);
+    LOG_DEBUG("crc\n");
+  }
+}
+
+static void prv_process_crc_rx(Fsm *fsm, const Event *e, void *context) {
+  CanDatagramStorage *store = context;
+  CanDatagram *dgram = &store->dgram;
+  if (!prv_read_rx_buffer((uint8_t*)&dgram->crc, CRC_SIZE_BYTES)) {
+    event_raise_no_data(store->repeat_event);
+  } else {
+    event_raise_no_data(store->rx_event);
+    LOG_DEBUG("dgram_type\n");
+  }
+}
+
+static void prv_process_dgram_type_rx(Fsm *fsm, const Event *e, void *context) {
+  CanDatagramStorage *store = context;
+  CanDatagram *dgram = &store->dgram;
+  if (!prv_read_rx_buffer(&dgram->dgram_type, DGRAM_TYPE_SIZE_BYTES)) {
+    event_raise_no_data(store->repeat_event);
+  } else {
+    event_raise_no_data(store->rx_event);
+    LOG_DEBUG("dst_len\n");
+  }
+}
+ 
+
+static void prv_process_dst_len_rx(Fsm *fsm, const Event *e, void *context) {
+  CanDatagramStorage *store = context;
+  CanDatagram *dgram = &store->dgram;
+  if (!prv_read_rx_buffer((uint8_t*)&dgram->destination_nodes_len, DEST_LEN_SIZE_BYTES)) {
+    event_raise_no_data(store->repeat_event);
+  } else {
+    event_raise_no_data(store->rx_event);
+    LOG_DEBUG("dst\n");
+  }
+}
+
+
+static void prv_process_dst_rx(Fsm *fsm, const Event *e, void *context) {
+  CanDatagramStorage *store = context;
+  CanDatagram *dgram = &store->dgram;
+//  size_t rx_len = (dgram->destination_nodes_len - store->rx_bytes_read) < CAN_BUFFER_SIZE ?
+//	  (dgram->destination_nodes_len - store->rx_bytes_read) : CAN_BUFFER_SIZE;
+//  if (!prv_read_rx_buffer((dgram->destination_nodes + store->rx_bytes_read), rx_len)) {
+//    event_raise_no_data(store->repeat_event);
+//  } else {
+//    store->rx_bytes_read += rx_len;
+//    LOG_DEBUG("RX LEN: %zu, %d\n", store->rx_bytes_read, dgram->destination_nodes_len);
+//    if (store->rx_bytes_read == dgram->destination_nodes_len) {
+//      LOG_DEBUG("data_len\n");
+//      event_raise_no_data(store->rx_event);
+//      store->rx_bytes_read = 0;
+//    } else {
+//      event_raise_no_data(store->repeat_event);
+//    }
+//  }
+  if (!prv_read_rx_buffer(dgram->destination_nodes, dgram->destination_nodes_len)) {
```

Comment:
@ryandancy (part 2) Because buffer is large enough to hold both the largest possible destination_nodes buffer (255) and  the largest possible data buffer (2048), we can read directly from the fifo in bulk, and eliminate a bunch of code.

I admit that this does seem a bit sus for `data`, as if it is the largest possible length it will fit the buffer exactly, but we can assume that for any given message, all other messages will be already read.

Tested it with max conditions and it is working, just wanted to clarify the idea here